### ISSUE TITLE
glTF support via Content Pipeline.

### ISSUE THREAD

**Issue by vpenades** [MonoGame/MonoGame#6932]
Currently, processing glTF files through the content pipeline is provided by Assimp importer, which presents a number of problems:
- Monogame Framework's Model miss a lot of glTF features, so when importing a glTF, many features are lost, specially those related to skeleton skinning, animation and materials.
- glTF is a relatively new format that keeps improving every day. Assimp is lagging behind into it's adoption, and MonoGame is lagging behind into adopting newer versions of Assimp, which results in a very outdated pipeline.

As a consequence, actual glTF support in monogame is very limited, when not broken.

In order to have a temporary solution, I developed a project that allows to load a glTF with full skeleton skinning and animation support, which can be found here:

~~https://github.com/vpenades/SharpGLTF/tree/master/examples/MonoGameScene~~
https://github.com/vpenades/SharpGLTF.Monogame.Example

But it still has two limitations:
- It uses MonoGame's BasicEffect and SkinnedEffect, so it's missing PBR materials.
- It needs to load the files at runtime, so it's limited to Desktop applications.

Ideally we should be able to process glTF files through the content pipeline with at least the same level of accuracy, or better.

So I've opened this discussion to talk about what might be needed in order to improve glTF support with the content pipeline.

### glTF features:

**File Specification (two variants)**

- glTF json files, which reference external binary blob files and texture files.
- GLB files which are self contained and have blobs and textures embedded.

**Layout specification**

- Default material is PBR, with additional materials available via extensions.
  - Texture formats:
    - JPEG
    - PNG
    - DDS (Extension)
    - WEBP (Extension)
    - Binomial Basis (Extension, WIP)
- An internal structure that treats meshes, textures, animations, etc as resources.
- Multiple scenes support.
- A single scene can contain:
  - One or many hierarchical node trees.
  - Cameras[Info]
  - Lights[Info]
  - Meshes
    - Rigid
    - Skinned
    - Morphable
    - Skinned + Morphable
    - Subdivision Surface Meshes (Extension, WIP)
- Nodes of the scene can have one or multiple animation tracks.
- Linear and Cubic spline curves are supported.
- Mesh GPU Instancing (Extension, WIP)
- LOD (Extension)

### Features required by MonoGame to fully support glTF

- The content pipeline only accepts model textures as external referenced files. So a new API would be required for embedded textures.
- We need a new PBR effect capable of handling skinning and morphing, with joint mapping and inverse bind matrices, hopefully, exceeding the 72 bone limit.
- A new Framework's Graphics Model class that covers all glTF features, and probably more.
- A new set of Content Pipeline's Builder classes able to handle glTF rich features.
- For animation, Quaternion curves are required.

**Comment by tomspilman** [comment]
For sure supporting a new importer would be great.

Also I don't mind see us extend the basic "model" support we have for something more modern.

But i don't want to see us have a "Scene", "Camera", or "Light" types of classes built into the core of MonoGame.  That takes us down the path of being an engine and starting to enforce rules on how a game scene has to work.

One possibility is to just have lightweight "CameraInfo" or "LightInfo" types and pass thru the raw data which can either be transformed by a custom game specific content processor or used raw at runtime for simple cases.  This lets the developer write their own camera and light systems.

Another possibility is to make all of this a completely separate assembly that is totally optional to include in your project if you want to use it.  Still we would want to keep all types flexible so they can be extended or modified without needing to dive into the source.

**Comment by vpenades** [comment]
@tomspilman I agree with your proposal of CameraInfo and LightInfo, actually, most engines just skip these nodes.

"Scenes" is an interesting thing, thought. I am using glTF naming convention, but in glTF a "Scene" is essentially the model's root node.

Although most glTF files you will encounter around always have a single Scene inside, the format specification actually supports multiple scenes, and there's a reason for it: to save memory on resources.

Internally, glTF treats meshes, textures, animations, etc as resources that can be instanced on every scene

Let's say you have 10 models that share not only most of the textures, but also lots of geometry meshes. In monogame, typically you would export each model separately, the textures would be shared, but the meshes and the animations would be duplicated across each model.

But with glTF, you would export a sort of "ModelCollection" with multiple Models inside, with the benefit that many MeshParts, textures and animations would be shared between meshes. Saving a lot of memory.

So far, I am not asking to support *all* the features of glTF, I am just showcasing these features to discuss what Monogame would support and what not.

**Comment by tomspilman** [comment]
> But with glTF, you would import a sort of "ModelCollection" with multiple Models inside

That seems reasonable to me.

**Comment by willmotil** [comment]
Last i looked the model class is just plain missing a lot even for fbx.
The loader i wrote out for assimp has a bunch of stuff that monogames importer and model class doesn't read in. Which is why i wrote my own model class to go with it. 
Im guessing that the other guys around here that wrote them have observed the same thing.

**Comment by vpenades** [comment]
Indeed, a prerequisite to support glTF, and more advanced FBX features, MonoGame would need a new Model class

**Comment by JoaoBaptMG** [comment]
The support for modern graphics in MonoGame is lagging entirely. No uniform/constant/shader storage buffers, no “advanced” texture types, no support for graphics/tessellation/compute shaders, or for unordered access views (aka image load/store). MonoGame is still pretty much in the pre-DX10 phase yet, and modern/PBR features require them, so I suggest waiting until proper support for modern techniques is ready before pushing such an advanced model.

**Comment by Jjagg** [comment]
> waiting until proper support for modern techniques is ready before pushing such an advanced model

If we sit around waiting nothing's gonna happen. The point of discussions like this is to make things happen. So I suggest we don't wait.

**Comment by JoaoBaptMG** [comment]
> So I suggest we don't wait.
We can use the proposal of supporting glTF to push the boundaries in the other directions, too.

**Comment by vpenades** [comment]
Then I believe the first thing to do would be to define an advanced Effect that supports all the required features; this requirement precedes anything else because all other glTF (or FBX features) will depend on this new advanced effect.

glTF uses [PBR Metallic Roughness](https://github.com/KhronosGroup/glTF/blob/master/specification/2.0/README.md#metallic-roughness-material) as default material, but by extension, other PBR formats are already supported, like [PBR Specular Glossiness](https://github.com/KhronosGroup/glTF/tree/master/extensions/2.0/Khronos/KHR_materials_pbrSpecularGlossiness) and more advanced ones are being developed.

Any glTF material supports, by default, these texture channels:
- Occlusion texture
- Normal texture
- Emissive texture

Metallic Roughness material adds these:
- BaseColor texture
- MetallicRoughness texture

Specular Glossiness material adds these:
- Diffuse texture
- Specular texture

Regarding vertex format support, glTF supports skinning with 4 and 8 joints, with unlimited number of joints. (to overcome the max bone limit, texture constants are used instead)

But to me, the most problematic feature is to support morph targets AND skinning, which ca be used at the same time, which requires to heave each morph target in an individual vertex buffer, and the vertex attributes in a separated vertex buffer, so you need to bind the appropiate morph target vertex buffers and the attributes vertex buffer simultaneously. I don't know if MonoGame can support this under the hood.

**Comment by JoaoBaptMG** [comment]
I’d say do away with the `Effect` class and somehow expose directly the shader state in the `GraphicsDevice`, along with some kind of vertex array object to store the attribute structure. I know Direct3D does not use a program-esque class anymore, and OpenGL already allows the creation of separate shader stages.

**Comment by tomspilman** [comment]
> I’d say do away with the Effect class and somehow expose directly the shader state in the `GraphicsDevice`,

We should support both so that people that want to write their own material stage can do so, but `Effect` exists as a starting point for those that don't.

We already have a `Shader` and `ConstantBuffer` API in there that `Effect` uses internally...  we just need to improve it and expose it.

**Comment by vpenades** [comment]
Regardless of whether  more or less API is exposed to allow developers to create more advanced effects/shaders, I believe it's about time we have an in-built Effect that goes beyond BasicEffect and it's up to date with modern technologies. Not everybody wants to deal with shaders, or reinvent the wheel for every new project...

**Comment by tomspilman** [comment]
> I believe it's about time we have an in-built Effect that goes beyond BasicEffect

For sure...  `PBREffect` would be a good start.

**Comment by willmotil** [comment]
Even before that what are we loading gltf with? Assimp isn't perfect at loading files either.
Last time i tried to load in gltf under Assimp it was a pain in the ass to get it to load. 
Much less properly make sense of it and Assimp doesn't always load all the material textures either even for fbx it also has some problems with scaling depending on what program exports the model and how it does so. For example blender exports with translational animations or bone animations but assimp makes no distinction between the two so you can't even detect it.

Deformations shouldn't even be that complicated its basically just more bones and vertice attributes.
Sounds like yet another crappy nonstandardized format in the making.

**Comment by vpenades** [comment]
> Even before that what are we loading gltf with? 

Assimp falls short to the task, indeed, so a glTF specific library will be needed.

One advantage of the content pipeline is that it can call to native libraries, and there's a number of glTF native libraries around.

If we want a 100% c# solution, we have [official khronos loader](https://github.com/KhronosGroup/glTF-CSharp-Loader) but it has not been updated for a while, and it's little more than a json schema to c# dump....

And then there's my [SharpGLTF](https://github.com/vpenades/SharpGLTF) library, which is what I am using in the [Runtime GLTF MonoGame Demo](https://github.com/vpenades/SharpGLTF/tree/master/examples/MonoGameScene) I showed in my first post.

**Comment by willmotil** [comment]
The ironic part here is that Assimp actually supports gltf and gltf2.0 as well as tons of other formats some that are actually easier to get working then fbx or .x ... But we only have the fbx and .x option in the content pipeline and the model itself ultimately is pretty easy to standardize. Once you get the data and the matrices in properly and can make sense of it.

**Comment by Jjagg** [comment]
> But we only have the fbx and .x option in the content pipeline.

We don't, we support everything assimp does.
Their glTF support is apparently not very good and of course the imported data must fit with their format which comes with a lot of issues, so I'd like us to use a glTF specific importer. I'm in favor of using SharpGLTF.

**Comment by Starnick** [comment]
FYI, AssimpNet has been updated to assimp 5.0 and is released in beta on nuget. From my understanding, there were many gltf fixes in 5.0

**Comment by vpenades** [comment]
> FYI, AssimpNet has been updated to assimp 5.0 and is released in beta on nuget. From my understanding, there were many gltf fixes in 5.0

And... Assimp.Net has been upgraded too? usually it lags behind the native counterpart...

But having better assimp.net support for gltf is only a small part of the problem, there's bigger ones: 

The monogame content pipeline requires textures to be in separate files, which prevents importing GLB files, from assimp or any other importer.

glTF's primary materials is PBR, which is missing by default in MonoGame.

And the biggest one: glTF has a rich animation and skinning system that cannot be replicated within monogame's runtime 'Model' counterpart, so a new, more advanced Model will be required.

So unless you're going to import non animated models with basic texture setups, you cannot import glTFs... and for doing just that there's no need for glTf, you can already go with wavefront obj and other good old similar formats.

**Comment by willmotil** [comment]
Ya its a pre-release im using it right now didn't try the gltf but the assimp site is doing what i would describe right now as house cleaning.  Gltf is still a work in progress over there.

Anyways if we use SharpGLTF just for gltf  and Assimp for the rest The model format has to cover pretty much the most complicated of both.  So i think its like.

GlTF > (GLTF) Content Pipeline> Monogame Model.
Assimp > (Not GLTF) Content Pipeline > Monogame Model.

> 
The monogame content pipeline requires textures to be in separate files, which prevents importing GLB files, from assimp or any other importer.

Maybe one of the other guys can comment on this it's probably possible to do it im not that good with it to say. But i was able to write a exporter importer that would combine files into one xnb.

_Well your gltf importer looks good i couldn't get it to run though when i gave it a shot but maybe i just gave up a bit to quickly_

**Comment by vpenades** [comment]
> Well your gltf importer looks good but i couldn't get it to work in mg but maybe i just gave up a bit to quickly

You mean this? https://github.com/vpenades/SharpGLTF/tree/master/examples/MonoGameScene

Which kind of issues did you have?

**Comment by Starnick** [comment]
> And... Assimp.Net has been upgraded too? usually it lags behind the native counterpart...

Yes thats what I mean. And yes it does lag, I know, I'm the author :)

**Comment by willmotil** [comment]
@vpenades 

> You mean this? https://github.com/vpenades/SharpGLTF/tree/master/examples/MonoGameScene
> Which kind of issues did you have?

Yes references errors mostly just didn't understand how to get it to build. 

I tried to download the whole thing and build it instead of cloning it, but the nuget references in many of the projects and other references were broken. 
I tried to just use the example projects in a separate solution and link in the nugets as per the notes on dependency's. 

> The project depends on SharpGLTF.Runtime.MonoGame, which also depends on SharpGLTF.Core, that's everything you need to load and render glTF models into MonoGame.

That didn't work either just more reference and nuget reference errors :(  
No instructions beyond that and no patience to guess at that many errors.

**Comment by vpenades** [comment]
@willmotil That's really weird...  anyway, I've just copied the monogame gltf demo to its own repository here: https://github.com/vpenades/SharpGLTF.Monogame.Example

It only references packages on github, so it should run out of the box: just clone, compile, and run MonoGameScene.exe.

Let me know if you had any problem with it.

**Comment by willmotil** [comment]
Excellent that works right off the rip no changes required at all.

vs2017 win 10.

**Comment by vpenades** [comment]
@willmotil the original repository has some net core 3 projects which require vs2019, so that's why it wasn't working for you.

glTF is designed around a very powerful skinning paradigm that gives a lot of flexibility to artists and allows for very complex scenes. But unfortunately this paradigm is not supported by most major engines, which struggle when importing some glTF files.

In fact, MonoGame does not support this paradigm in its `Model` class either. That's why I had to create a new custom model, and why I am proposing a new model object for monogame

**Comment by willmotil** [comment]
@vpenades i didn't have time to take much of a look at it yesterday but i was glancing it over today. 
Im really impressed. Not that i understand a bit of it wish i did.

So just to run a futher test.
I went onto sketch fab found and downloaded a animated fbx that was auto converted to gltf. 
I then looked up another converter, to convert it to a .glb. 
I was suprised to find a drag and drop web converter here.
https://sbtron.github.io/makeglb/

I replaced one of the models you had loaded with that twice converted model and ran it.

![gltfconvertedmodeltest02](https://user-images.githubusercontent.com/7495196/75116325-2dcebb80-5635-11ea-8650-8529cb9142ad.gif)
